src/ems_service.ts

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { EMSClient, BaseEmsServiceConfig } from './ems_client';
+import { ORIGIN } from './origin';
+import { toAbsoluteUrl } from './utils';
+
+export interface IEmsService {
+  getAttributions(): { url: string; label: string }[];
+  getHTMLAttribution(): string;
+  getMarkdownAttribution(): string;
+  getDisplayName(): string;
+  getId(): string;
+  hasId(id: string): boolean;
+  getOrigin(): string;
+  getApiUrl(): string;
+}
+
+export class AbstractEmsService implements IEmsService {
```

Comment:
you can mark this class as [abstract](https://www.typescriptlang.org/docs/handbook/classes.html#abstract-classes) and later down mark the methods you are not really implementing here so you are sure your inherited classes have them.

```typescript
  abstract getDisplayName(): string;

  abstract getId(): string;

  abstract hasId(id: string): boolean;
  
  abstract getApiUrl(): string;
```

Reply:
ooh! thanks!